src/sat/tactic/sat_tactic.cpp

```diff
@@ -29,7 +30,7 @@ class sat_tactic : public tactic {
         ast_manager &   m;
         goal2sat        m_goal2sat;
         sat2goal        m_sat2goal;
-        scoped_ptr<sat::solver_core>   m_solver;
+        scoped_ptr<sat::solver> m_solver;
```

Comment:
this was introduced at a point when Maate Soos was considering integrating CMS as a secondary backend. Having an abstraction layer for the solver would tell what features of a SAT solver to support. This abstraction layer would support minimal features and therefore not extensions. 
Since the CMS experiment never happened, it is very possible we should just abandon sat_solver_core.

Reply:
Is the CMS experiment still “interesting”? If so, I’d be happy to pick-up where Mate left.

Reply:
Maate didn't "go" anywhere. There was a collaborator who considered it, but it is about +2 years ago.

Note that it will be essentially impossible to use CMS for the SMT extensions (and I don't see why it would be useful).

Also, the starting point is _not_ to release z3 with CMS dependencies. It will tax development time and hurt more users who don't need this.

Something to consider is also that plugging in CMS just at the level of the SAT solver might also be the wrong design: having a self-contained non-contaminating pipeline or even a self-contains cms_tactic could be the way to go. Then it would have to do its own CNF transformation but have the benefit of being "one-shot" so no complications around incrementality. Of course there can be a use case for incrementality so the tactic route is painting into a corner.

It is really a question of what would it be used for.

Reply:
> Also, the starting point is not to release z3 with CMS dependencies.
> It will tax development time and hurt more users who don't need
> this.
>

100% agree -- if I were to pick this up, it would be an CMake-only "option" (e.g., `-DENABLE_CMS:BOOL=ON`) that would only work if CMake could find CMS on the system.

If CMake can't find CMS, or you don't ask for CMS, then Z3 would be "unchanged".

> that it will be essentially impossible to use CMS for the SMT
> extensions
>

Interesting; I guess I don't understand enough of the Z3 archictecture here; I was thinking that CMS could be "plugged in" as part of a typical DPLL(T)-loop.


> plugging in CMS just at the level of the SAT solver might also be
> the wrong design
>

Would it be at all interesting in doing this as a "Step 1" just to show viability? It seems that this would be the lowest-hanging fruit, even if it wasn't the end-goal in itself.

Basically, taking something too ambitious to start with ends-up with stagnation in my experience.

Reply:
The SMT integration creates and deletes variables during search. So "impossible" means "a lot of extra work".

Reply:
> it would be an CMake-only "option" 

probably, it would also probably bit-rot at some time and then be kicked out when not maintained.
z3 currently has no "links" to other github projects (commonly used elsewhere). It is also a feature based on (1) my ignorance of using features (2) keeping the build process as friction less as possible (it is still not without friction).
Of course, there are various plugin dependencies for API languages (OCaml, .Net, Java), and the main grief is that not everything builds on all variants. There are many complaints about OCaml issues. Nevertheless OCaml is being used.
Just yesterday I learned about a rather interesting and useful application: https://engineering.fb.com/2021/09/13/core-data/superpack/